### ISSUE TITLE
Pass selected elements instead of resampling

### ISSUE THREAD

**Issue by moustachio-belvedere** [JuliaRheology/RHEOS.jl#14]
Sometimes if we want the model fitting to focus on particularly troublesome areas we resample with higher sample rates in the difficult-to-fit areas.

However, this leads to a variable sample rate which forces the use of trapezoidal computation of the hereditary integral which is very slow.

Instead, we should have another downsample-only function which returns either the element numbers desired or a boolean mask array (with desired elements set to true). 

This could then be passed to the objective function such that the only elements considered by the optimizer are the desired ones.

If sample rate is already low, could do a fixed upsample, followed by the above process.

In most cases it will be **MUCH** faster.

**Comment by moustachio-belvedere** [comment]
Consider the following code example:

```
julia> t = collect(0.0:0.5:3.0);
julia> desired_els = [true, true, true, false, true, false, true];
julia> t[desired_els]
5-element Array{Float64,1}:
 0.0
 0.5
 1.0
 2.0
 3.0
```
Or equivalently 
```
julia> t = collect(0.0:0.5:3.0);
julia> desired_els = [1, 2, 3, 5, 7];
julia> t[desired_els]
5-element Array{Float64,1}:
 0.0
 0.5
 1.0
 2.0
 3.0
```
Either way, we can just pass `desired_els` to the objective function and it will work nicely.

**Comment by moustachio-belvedere** [comment]
see this old version of downsample (from julia 0.6 branch version of RHEOS:
```

function downsample(boundaries::Array{Int64,1}, elperiods::Array{Int64,1})

    # assert correct function signature
    @assert length(elperiods)==length(boundaries)-1 "Number of different sample periods must be 1 less than boundaries provided"

    # initialise indices array
    indices = zeros(Int64,0)

    # loop through, skipping elements as required
    for i in 1:length(boundaries)-1

        # get indices for this 'section'
        indicesCur = boundaries[i]:elperiods[i]:boundaries[i+1]

        # append
        append!(indices, collect(indicesCur))
    end

    # return unique indices
    indices = unique(indices)
end
```

These indices could be passed to the cost function e.g.
```
function cost(...)
    compute_hereditary_integral()

    return sum( ((compute_hereditary_integral .- measured).^2)[desired_indices] )
end
```

So only desired indices would affect the optimisation. **_Note that in the above, `compute_hereditary_integral would be the constant sample rate version, which is much faster than the variable sample rate version._**

**Comment by akabla** [comment]
couldn't we use resample / cutting already for this?

**Comment by moustachio-belvedere** [comment]
`cutting` would be fine if you just wanted a single subsection of the data to be fitted.

What about if you want the cost contribution of every element from 0<t<10 and then every third element form 10<t<20 and every fifth element for the rest of the time sample? This is a very real use case, Ale and I have both done similar things manually at some point.

At the moment you could `resample` and disgard extraneous datapoints before you start fitting, but then you have a variable sample rate. This will be very slow because RHEOS will be forced to use the trapezoidal computation of the hereditary integral O(N^2) which requires O(N^2) calls to the Mittag-Leffler function.

If you just pass the indices which the optimiser should specialise on then the data can retain its constant sample rate and the hereditary integral is computed via FFT convolution which is O(nlogn), with only O(n) calls to the Mittag-Leffler function.  

In other words, the 'downsampling' only happens after the hereditary intragral has been computed.

**Comment by akabla** [comment]
Ah, understood! this sounds like a good idea indeed, as long as it doesn't
slow down normal use.

**Comment by moustachio-belvedere** [comment]
Yes good point, I think to 'resolve' this issue we should also benchmark to make sure passing indices doesn't add significant overhead to the cost function. 

If it does we can outsource to specialised cost functions so that there is no overhead added when **_not_** passing indices.

**Comment by moustachio-belvedere** [comment]
Benchmarked in the file: https://github.com/JuliaRheology/RheoBenchAndTest/blob/master/20190924_selectedindicesCostfunction.jl

Results are pretty consistent:

```
julia> include("20190924_selectedindicesCostfunction.jl")
  11.428 ms (171 allocations: 16.95 MiB)
  11.909 ms (174 allocations: 17.71 MiB)
```

Difference is only slight but will do two separate cost functions as per discussion with Alexandre.

Also, name indices function something like `indicesweighting` which can "downsample" by not including every element and "upsample" by repeating elements and thus weighting them higher.

**Comment by moustachio-belvedere** [comment]
`indexweight` function added in commit b63dce650b004923c535c6d471967a7bf04d1dbf
`indexweight`  tests added in commit 5d56bb4279bafd50ebafea9efbebe5b226172ce5
`indexweight` docstring improved in commit 8056b59d2469b36c562256a2f99b620a160da016

Now just need to add the two objective functions with indices selection and modify `modelfit` slightly.

**Comment by moustachio-belvedere** [comment]
Only implementing for constant sample-rate data. `modelfit` and `modelstepfit` will throw warnings when used with weighted indices and variable sample-rate for now. This is so there is only one weighting pathway for constant-sample-rate data (weighted indices) and one weighting pathway for variable-sample-rate data (resampling).

**Comment by moustachio-belvedere** [comment]
`modelfit` is done in commits
fdde4b811468dcac8c6f17fc3f91e1a25283d237
e46f5dcc8a3820a0ed270e7e91105b2fba34147b

**Comment by moustachio-belvedere** [comment]
and `modelstepfit` in commit 4325908c45855fc01b9cf83bc47a6ae9a5b85d26

Done for now. Will add for differential equations when resolving #54
